### ISSUE TITLE
Reverse join not saving

### ISSUE THREAD

**Issue by bedengler** [atk4/data#799]
Following situation:

Table 1 (contact_table): 
contact_id (primary), tbl1_name, tbl1_address

Table_2 (additional_info_table): 
additional_id (primary), additional_contact_id (relation), additional_content

Let's assume I have a contact table which is loaded always (Table 1). 
I want to read, if there's an entry available from an additional table (Table 2). Table 2 always references to Table 1.

```
class myModel extends \atk4\data\Model {
    use \atk4\core\TrackableTrait;
    use \atk4\core\AppScopeTrait;

    public $table = "contact_table";
    public $id_field = "contact_id";
    
    function init(): void {
        parent::init();

        $this->leftJoin($this->'additional_info_table', [
          'master_field' => 'contact_id',
          'foreign_field' => 'additional_contact_id',
          'reverse' => true
        ]);
    }
}
```

When using this model in a form, it reads properly, but it just saves data from the main table. The data of the joined table is neither inserted nor updated.
Is this how it's intended to work? If yes, then how to deal with a situation like this?

**Comment by bedengler** [comment]
Here's an example sql file with 2 tables adapted to above described scenario...
[join-example-table_2020-12-08.sql.zip](https://github.com/atk4/data/files/5660517/join-example-table_2020-12-08.sql.zip)

**Comment by bedengler** [comment]
This is working, but just if there exists an entry in the "additional_info_table", otherwise the result is completely empty:
```
$join = $this->join($this-> additional_info_table, [
          'on'=>$this->expr('contact_id = _st. additional_contact_id'),
          'foreign_alias'=>'_st',
          'master_field' => 'contact_id',
          'foreign_field' => 'additional_contact_id',
        ]);
```

**Comment by DarkSide666** [comment]
What is `$this->'additional_info_table'` ? That's wrong. Should be just table name there without $this.

**Comment by bedengler** [comment]
Ah sorry, I'm calling the function with a variable that's called like this...
So for the example it was missleading but in the real world application it works...

It's like
```
class  .... {

$additional_info_table = "additional_table";

function ....... {
    $join = $this->join($this->additional_info_table....
....
```

But you are absolutely right, $this->additional_info_table could be replaced with 'additional_table'

**Comment by DarkSide666** [comment]
First of all you have to:
1. understand what type of relationship it is. Is it 1:1 or 1:n? 
2. I don't see any _reverse_ join situation here.
3. If relation is 1:n then your myModel will contain more than one record for one contact if you join multiple additional_contact_info records.
4. is it weak join (left join) or full join? It will save info in joined table only in case of full join.

So as result I think you only need this:
```
    $j = $this->leftJoin($this->additional_info_table, [
        'master_field' => 'contact_id',
        'foreign_field' => 'additional_contact_id',
    ]);
    $j->addField(....);
```

**Comment by bedengler** [comment]
It's a 1:1 and it doesn't need the "on" I realised in the meanwhile.
Yes, I need full join and to save / add an entry. 
This...
```
$join = $this->join($this-> additional_info_table, [
          'foreign_alias'=>'_st',
          'master_field' => 'contact_id',
          'foreign_field' => 'additional_contact_id',
        ]);
```
works as well...BUT:

I have loaded the model with join on a CRUD. Works basically until I try to make a new entry. It doesn't save the entry at all and the error message says: "atk4\data\Exception [code: 404]: Record was not found " which is very strange...and the auto_increment counter counts +1 (of the contact_table - Table 1), but there's no entry in it.

As soon as I deactivate the join everything works again.

So basically everything works except adding a new entry...

**Comment by DarkSide666** [comment]
That's strange. Do you add some fields from joined table in model too?

Asking because one thing I can imagine is that it actually tried to insert record in parent table, but for some reason didn't insert respective record in child table. Maybe because you have not added a single field from child table into model. As result when model is trying to load back this newly inserted record it's not found because there is no record in full join table. And as result you get exception and rollback and for some reason rollback is not affecting auto increment value.

Can you provide all your model setup (with joins and all addFields) and also DB table definitions with null/not null and constraints?

**Comment by bedengler** [comment]
Yes, I'm also adding fields from the joined table.
Ok, here is my full code that produces this behaviour incl. database. I hope you can work with that!?

[joinTest.zip](https://github.com/atk4/data/files/5668957/joinTest.zip)

I'm currently experimenting with a workaround using a hook. 
Every time the main model gets an insert, the hook makes a new entry in the additional table. 
But of course it would be more elegant, if the join would do everything automagically... 😉

**Comment by bedengler** [comment]
This does the trick, although it's less elegant.
I put it within AdspacesAdDetailAllModel
```
$this->onHook(
          \atk4\data\Model::HOOK_AFTER_INSERT,
          function ($x, $insertedID) {
            $addon = new AdspacesAdDetailAddonModel($this->persistence);  // Model of the additional table
            $addon->set('adaddon_Ad_ID', $insertedID);
            $addon->save();
          }
        );
```
Now everything works as expected...everytime when I'm generating a new entry, the other table gets also a new entry

**Comment by DarkSide666** [comment]
Yes with a hook it works, but that's what atk4/data should do out of the box and it did that nicely some time ago.

**Comment by DarkSide666** [comment]
Ok checking. Looks like it's really a bug in atk4/data.

Edit: It's not exactly a bug, but a bit strange implementation of reverse joins indeed.
On another note if it's intended to be 1:1 relationship between these tables, then your DB structure looks a bit wrong. You should better move joining id field to another (parent) table. That would really make sure you're having 1:1 not 1:n as it is implemented currently.

**Comment by DarkSide666** [comment]
So this was not 100% like a bug, but still it had some space for improvement (done in https://github.com/atk4/data/pull/802).

To generate reverse join (which you need in your DB structure case) you have to join `table.field`.
Another option is to use `reverse=>true` setting, but that didn't work well before #802.

So in your specific example you have to:
```
public $table_to_join = 'megaboard985_adspaces_addon.adaddon_Ad_ID'; // <-- add foreign field name here !!!

// and no need to set fields or reverse here at all, that will be done automatically
$join = $this->join($this->table_to_join);
```

after #802 you can also define it like this with same results:
```
public $table_to_join = 'megaboard985_adspaces_addon';

// and no need to set fields or reverse here at all, that will be done automatically
$join = $this->join($this->table_to_join, [
    'reverse' => true,
    'foreign_field' => 'adaddon_Ad_ID',
]);
```

P.S. Few things to improve:
- your table and field names quite are terrible :) So hard to pronounce and remember (`megaboard985_adspaces_addon.adaddon_Ad_ID`)
- it would be much easier for you if you don't use mixed case letters in table and field names in DB and in ATK
- in case of 1:1 relation consider changing DB structure and moving adaddon_Ad_ID field to parent (mam_adspaces) table so it works without reverse join magic and would be much more straightforward.

**Comment by bedengler** [comment]
Ok thanks for all your effort and input. 🙏
I think you are right concerning the 1:1 connection. I'll think about moving the relation field.

And thanks for fixing. 😉👍

Concerning column and table names, I have a naming convention for that:
+ Every field name must be unique and self-explaining
+ Every table has a prefix (e.g. "adaddon_") that's part of each field name
+ Table names are all over lower case
+ Field names are first letter lower case, the rest has first letter in uppercase
+ "ID" is always uppercase

In my logic this makes it easy to distinguish between:
+ Tables
+ Fields without ID function
+ Fields with ID (which are mostly referring to other tables)

Additionally I don't have to care about conflicts (e.g. mixing "id" of several tables on joins etc.).

Having this in mind makes it (for me) pretty straightforward to navigate the database... 🤗
